docs/decisions/0016-python-public-api-conventions.rst

```diff
@@ -0,0 +1,60 @@
+
+For these reasons, we should supplement OEP-49 recommendations with a more consumer-friendly API layer.
+
+Decision
+--------
+
+In addition to OEP-49, we will adopt the following practices:
+
+Learning Core Django apps will be grouped into packages.
+  Apps in ``openedx_learning`` will be grouped into broadly related packages under ``openedx_learning.apps``. The first of these groups will be "authoring" (``openedx_learning.apps.authoring``). Future packages may include "learner", "personalization", "activity", "grading", etc.
+
+Learning Core Django apps will continue to have their own ``api`` modules.
+  So for example, ``openedx_learning.apps.authoring.components.api`` will continue to exist.
+
+Learning Core will have a top level package for its public API.
+  All public APIs intended for use by consumers of Learning Core will be represented as modules in the ``openedx_learning.api`` package that corresponds to the app groupings (e.g. ``openedx_learning.api.authoring``).
+
+App ``api`` modules will define their public functions using ``__all__``.
+  The public API modules will do a wildcard import from the various apps in their package group. So ``openedx_learning/api/authoring.py`` might look like::
+
+    from ..apps.authoring.components.api import *
+    from ..apps.authoring.contents.api import *
+    from ..apps.authoring.publishing.api import *
+
+  This relies on the individual apps to properly set ``__all__`` to the list of functions that they are willing to publicly support.
+
+App ``api`` modules within a package of apps still import from each other.
+  So for example, ``openedx_learning.apps.authoring.components.api`` will continue to import APIs that it needs from ``..publishing.api``, instead of using the public API at ``openedx_learning.api.authoring``. These imports should not use wildcards.
+
+  Functions and constants that are not listed as part of a module's ``__all__`` may still be imported by other app APIs in the same package grouping. This should allow a package more flexibility to create provisional APIs that we may not want to support publicly.
+
+  If a function or attribute is intended to be completely private to an app's ``api`` module (i.e. not used even by other apps in its package), it should be prefixed with an underscore.
+
+App ``api`` modules should not import directly from apps outside their package.
+  For example, ``openedx_learning.apps.personalization.api`` should import authoring API functions from ``openedx_learning.api.authoring``, **not** directly from something like ``openedx_learning.apps.authoring.components.api``. This will help to limit the impact of refactoring app package internal changes, as well as exposing shortcomings in the existing public APIs.
+
+Public API modules may implement their own functions.
```

Comment:
I like this. Not only for convenience but also as a way to address the natural evolution of the API. Do you think this might replace versioning by creating shims and deprecating functions with good documentation and timelines?

Reply:
I'm not sure if it will replace versioning altogether, but I definitely plan to use compatibility shims more liberally here than we have in for Django apps historically, and to make deprecations happen at the level of individual functions with deprecation warnings.